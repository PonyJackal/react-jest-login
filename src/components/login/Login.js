import React, { useState } from 'react'
import './Login.scss'

const Login = () => {
    const [formData, setFormData] = useState({
        values: {
            email: '',
            password: ''
        },
        errors: {
            email: '',
            password: ''
        },
        isValid: false
    })

    const handleChange = (event) => {
        setFormData({
            ...formData,
            values: {
                [event.target.name]: event.target.value
            }
        })
    }
    const handleSubmit = () => {
        if(!formData.isValid){
            console.log(formData.values)
        }
    }

    return (
        <div className="login">
            <h1>Login Form</h1>
            <form className="login-form">
                <label>Email: <input name="email" onChange={handleChange} vlaue={formData.values.email} /></label>
                <label>Password: <input name="password" onChange={handleChange} vlaue={formData.values.password}/></label>
                <button onClick={handleSubmit} disabled={!formData.isValid}>Login</button>
            </form>
        </div>
    )
}

export default Login