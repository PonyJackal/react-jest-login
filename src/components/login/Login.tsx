import React, { useState } from 'react'
import './Login.scss'

const Login = () => {
  //sate for form
  const [formData, setFormData] = useState({
    values: {
      email: '',
      password: ''
    },
    errors: {
      email: '',
      password: ''
    },
    isValid: false,
    isSubmittted: false
  })
  //input change handler
  const handleChange = (event) => {
    const { name, value } = event.target
    setFormData({
      ...formData,
      values: {
        ...formData.values,
        [name]: value
      }
    })
  }
  //form submission handler
  const handleSubmit = (event) => {
    const errors = validation()
    const isValid = !errors.email && !errors.password ? true : false
    setFormData({
      ...formData,
      errors: {
        email: errors.email,
        password: errors.password
      },
      isValid: isValid
    })
    if (isValid) {
      setFormData({
        ...formData,
        isSubmittted: true,
        errors: {
          email: "",
          password: ""
        }
      })
    }
  }
  //form validation
  const validation = () => {
    const regex = /^[^\s@]+@[^\s@]+\.[^\s@]{2,}$/i
    const errors = {
      email: "",
      password: ""
    }

    if (!formData.values.email) {
      errors.email = "Email cannot be blank"
    }
    else if (!regex.test(formData.values.email)) {
      errors.email = "Invalid email format"
    }

    if (!formData.values.password) {
      errors.password = "Password cannot be blank"
    }
    else if (formData.values.password.length < 4) {
      errors.password = "Password must be more than 4 characters"
    }
    return errors;
  }

  return (
    <div className="container">
      <h1>Sign in to continue</h1>
      {formData.isSubmittted && (
        <span className="success-msg">Form submitted successfully</span>
      )}
      <form>
        <div className="form-row">
          <label htmlFor="email">Email</label>
          <input
            type="email"
            name="email"
            id="email"
            value={formData.values.email}
            onChange={handleChange}
            className={formData.errors.email && "input-error"}
          />
          {formData.errors.email && (
            <span className="error">{formData.errors.email}</span>
          )}
        </div>

        <div className="form-row">
          <label htmlFor="password">Password</label>
          <input
            type="password"
            name="password"
            id="password"
            value={formData.values.password}
            onChange={handleChange}
            className={formData.errors.password && "input-error"}
          />
          {formData.errors.password && (
            <span className="error">{formData.errors.password}</span>
          )}
        </div>

        <button data-testid="submit-button" type="button" onClick={handleSubmit}>Sign In</button>
      </form>
    </div>
  )
}

export default Login