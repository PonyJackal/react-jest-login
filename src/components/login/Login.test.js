import React from 'react'
import ReactDOM from 'react-dom'
import { render, screen, cleanup, fireEvent } from '@testing-library/react'
import renderer from 'react-test-renderer'

import Login from './../login/Login'

afterEach(cleanup)

describe("render components without crashing", () => {
    it("renders without crashing", () => {
        const div = document.createElement('div')
        ReactDOM.render(<Login />, div)
    })

    it("matches snapshot", () => {
        const login = renderer.create(<Login />).toJSON()
        expect(login).toMatchSnapshot()
    })
})

describe("check validation", () => {
    it("Email can not be blank", () => {
        const login = render(<Login />)
        const inputEmail = login.getByLabelText("Email")
        const submitButton = login.getByTestId('submit-button')

        fireEvent.change(inputEmail, { target: { value: "" } })
        fireEvent.click(submitButton)

        screen.getByText("Email cannot be blank")
    })

    it("Check invalid email", () => {
        const login = render(<Login />)
        const inputEmail = login.getByLabelText("Email")
        const submitButton = login.getByTestId('submit-button')

        fireEvent.change(inputEmail, { target: { value: "adsf" } })
        fireEvent.click(submitButton)

        screen.getByText("Invalid email format")
    })

    it("Password can not be blank", () => {
        const login = render(<Login />)
        const inputPassword = login.getByLabelText("Password")
        const submitButton = login.getByTestId('submit-button')

        fireEvent.change(inputPassword, { target: { value: "" } })
        fireEvent.click(submitButton)

        screen.getByText("Password cannot be blank")
    })

    it("Password must be more than 4 characters", () => {
        const login = render(<Login />)
        const inputPassword = login.getByLabelText("Password")
        const submitButton = login.getByTestId('submit-button')

        fireEvent.change(inputPassword, { target: { value: "aa" } })
        fireEvent.click(submitButton)

        screen.getByText("Password must be more than 4 characters")
    })

    it("Form Success", () => {
        const login = render(<Login />)
        const inputEmail = login.getByLabelText("Email")
        const inputPassword = login.getByLabelText("Password")
        const submitButton = login.getByTestId('submit-button')

        fireEvent.change(inputEmail, { target: { value: "a@a.com" } })
        fireEvent.change(inputPassword, { target: { value: "1234" } })
        fireEvent.click(submitButton)

        screen.getByText("Form submitted successfully")
    })
})
