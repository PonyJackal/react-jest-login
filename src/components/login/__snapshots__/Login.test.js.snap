// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render components without crashing matches snapshot 1`] = `
<div
  className="container"
>
  <h1>
    Sign in to continue
  </h1>
  <form>
    <div
      className="form-row"
    >
      <label
        htmlFor="email"
      >
        Email
      </label>
      <input
        className=""
        id="email"
        name="email"
        onChange={[Function]}
        type="email"
        value=""
      />
      
    </div>
    <div
      className="form-row"
    >
      <label
        htmlFor="password"
      >
        Password
      </label>
      <input
        className=""
        id="password"
        name="password"
        onChange={[Function]}
        type="password"
        value=""
      />
      
    </div>
    <button
      data-testid="submit-button"
      onClick={[Function]}
      type="button"
    >
      Sign In
    </button>
  </form>
</div>
`;
